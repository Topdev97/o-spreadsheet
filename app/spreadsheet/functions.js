export const functions = {
  SUM: {
    description: "Returns the sum of all values in a range",
    compute: function(range) {
      return range.reduce((a, b) => a + b, 0);
    }
  }
};
