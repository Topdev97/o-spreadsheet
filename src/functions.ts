export const functions = {
  SUM: {
    description: "Returns the sum of all values in a range",
    compute: function(...args) {
      return args.flat().reduce((a, b) => a + b, 0);
    }
  },
  RAND: {
    description: "Returns a random number between 0 and 1",
    compute: function() {
      return Math.random();
    }
  }
};
