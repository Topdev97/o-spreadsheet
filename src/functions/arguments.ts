export type ArgType = "BOOLEAN" | "ANY" | "RANGE" | "NUMBER" | "STRING";

export interface Arg {
  repeating?: boolean;
  optional?: boolean;
  description: string;
  name: string;
  type: ArgType[];
  default?: any;
  valueProvider?: Function;
  isValueProviderRestrictive?: boolean;
}

//------------------------------------------------------------------------------
// Argument validation
//------------------------------------------------------------------------------

export function validateArguments(args: Arg[]) {
  let previousArgRepeating: boolean | undefined = false;
  let previousArgOptional: boolean | undefined = false;
  for (let current of args) {
    if (previousArgRepeating) {
      throw new Error(
        "Function ${name} has at least 2 arguments that are repeating. The maximum repeating arguments is 1."
      );
    }

    if (previousArgOptional && !current.optional) {
      throw new Error(
        "Function ${name} has at mandatory arguments declared after optional ones. All optional arguments must be after all mandatory arguments."
      );
    }
    previousArgRepeating = current.repeating;
    previousArgOptional = current.optional;
  }
}

//------------------------------------------------------------------------------
// Wrapping functions for arguments sanitization
//------------------------------------------------------------------------------
export function protectFunction(fn: Function, argList: Arg[]): Function {
  if (argList.length === 0) {
    return fn;
  }
  return function(this: any, ...args: any[]) {
    args = sanitizeArgs(args, argList);
    return fn.call(this, ...args);
  };
}

/**
 * If you read this, and are horrified by the code, worry not, dear friend.
 * This code is here only for a while, to solidify the test suite and prepare
 * the future. It will be replaced by a shiny argument sanitizer compiler soon.
 *
 * Note: this function modifies args in place!
 */
export function sanitizeArgs(args: any[], argList: Arg[]): any[] {
  for (let i = 0; i < argList.length; i++) {
    const descr = argList[i];
    if (!(i in args) && !descr.optional) {
      throw new Error("Wrong number of arguments. Expected 1, but got 0 argument instead.");
    }
    const arg = args[i];
    if (arg === undefined && descr.optional) {
      args = args.slice(0, i);
      break;
    }
    if (descr.repeating) {
      for (let j = i; j < args.length; j++) {
        sanitizeArg(args, j, args[j], descr);
      }
      return args;
    }
    sanitizeArg(args, i, arg, descr);
  }
  if (args.length > argList.length) {
    throw new Error(
      `Wrong number of arguments. Expected ${argList.length}, but got ${args.length} arguments instead.`
    );
  }
  return args;
}

function sanitizeArg(args: any[], i: number, arg: any, descr: Arg) {
  if (arg === undefined) {
    if (descr.type.includes("NUMBER")) {
      args[i] = 0;
    }
    if (descr.type.includes("BOOLEAN")) {
      args[i] = false;
    }
  } else if (typeof arg === "boolean" && !descr.type.includes("BOOLEAN")) {
    if (descr.type.includes("NUMBER")) {
      args[i] = arg ? 1 : 0;
    }
  } else if (typeof arg === "string") {
    if (descr.type.includes("NUMBER")) {
      if (arg) {
        const n = Number(arg);
        if (isNaN(n)) {
          throw new Error(
            `Argument "${descr.name}" should be a number, but "${arg}" is a text, and cannot be coerced to a number.`
          );
        }
      } else {
        args[i] = 0;
      }
      args[i] = arg ? parseFloat(arg) : 0;
    } else if (descr.type.includes("BOOLEAN")) {
      if (arg === "") {
        args[i] = false;
      } else if (arg.toUpperCase() === "TRUE") {
        args[i] = true;
      } else if (arg.toUpperCase() === "FALSE") {
        args[i] = false;
      } else {
        throw new Error(
          `Argument "${descr.name}" should be a boolean, but "${arg}" is a text, and cannot be coerced to a boolean.`
        );
      }
    }
  } else if (typeof arg === "number") {
    if (descr.type.includes("BOOLEAN")) {
      args[i] = arg ? true : false;
    }
  }
}
//------------------------------------------------------------------------------
// Arg description DSL
//------------------------------------------------------------------------------

const ARG_REGEXP = /(.*)\((.*)\)(.*)/;
const ARG_TYPES: ArgType[] = ["BOOLEAN", "NUMBER", "STRING", "ANY", "RANGE"];

/**
 * This function is meant to be used as a tag for a template strings.
 *
 * Its job is to convert a textual description of the list of arguments into an
 * actual array of Arg, suitable for consumption.
 */
export function args(strings: TemplateStringsArray): Arg[] {
  let lines = strings[0].split("\n");
  const result: Arg[] = [];
  for (let l of lines) {
    l = l.trim();
    if (l) {
      result.push(makeArg(l));
    }
  }
  return result;
}

function makeArg(str: string): Arg {
  let parts = str.match(ARG_REGEXP)!;
  let name = parts[1].trim();
  let types: ArgType[] = [];
  let isOptional = false;
  let isRepeating = false;
  for (let param of parts[2].split(",")) {
    param = param.trim().toUpperCase();
    let type = ARG_TYPES.find(t => param === t);
    if (type) {
      types.push(type);
    } else if (param === "OPTIONAL") {
      isOptional = true;
    } else if (param === "REPEATING") {
      isRepeating = true;
    }
  }
  let description = parts[3].trim();
  const result: Arg = {
    name,
    description,
    type: types
  };
  if (isOptional) {
    result.optional = true;
  }
  if (isRepeating) {
    result.repeating = true;
  }
  return result;
}
