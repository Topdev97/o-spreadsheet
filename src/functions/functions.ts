//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------
export interface FunctionDescription {
  description: string;
  compute: Function;
  async?: boolean;
  category?: string;
  args: Arg[];
  returns: [ArgType];
}

export type FunctionMap = { [key: string]: FunctionDescription };
export type ArgType = "BOOLEAN" | "ANY" | "RANGE" | "CELL" | "NUMBER" | "STRING";

export interface Arg {
  repeating?: boolean;
  optional?: boolean;
  description: string;
  name: string;
  type: ArgType[];
  default?: any;
  valueProvider?: Function;
  isValueProviderRestrictive?: boolean;
}

//------------------------------------------------------------------------------
// Functions
//------------------------------------------------------------------------------

// todo: make name more descriptive, and add some docstring

/**
 * Mapping from function names to descriptions
 */
export const functions: FunctionMap = {};

/**
 * Mapping from function name to the corresponding compute method
 */
export const functionMap: { [name: string]: Function } = {};

//------------------------------------------------------------------------------
// Others
//------------------------------------------------------------------------------

function validateArguments(args: Arg[]) {
  let previousArgRepeating: boolean | undefined = false;
  let previousArgOptional: boolean | undefined = false;
  for (let current of args) {
    if (previousArgRepeating) {
      throw new Error(
        "Function ${name} has at least 2 arguments that are repeating. The maximum repeating arguments is 1."
      );
    }

    if (previousArgOptional && !current.optional) {
      throw new Error(
        "Function ${name} has at mandatory arguments declared after optional ones. All optional arguments must be after all mandatory arguments."
      );
    }
    previousArgRepeating = current.repeating;
    previousArgOptional = current.optional;
  }
}

/**
 * Add a function to the internal function list.
 */
export function addFunction(name: string, descr: FunctionDescription) {
  name = name.toUpperCase();
  if (name in functionMap) {
    throw new Error(`Function ${name} already registered...`);
  }

  validateArguments(descr.args);

  functionMap[name] = descr.compute;
  functions[name] = descr;
}

//------------------------------------------------------------------------------
// Arg description DSL
//------------------------------------------------------------------------------

const ARG_REGEXP = /(.*)\((.*)\)(.*)/;
const ARG_TYPES: ArgType[] = ["BOOLEAN", "NUMBER", "ANY", "RANGE", "CELL", "NUMBER", "STRING"];

/**
 * This function is meant to be used as a tag for a template strings.
 *
 * Its job is to convert a textual description of the list of arguments into an
 * actual array of Arg, suitable for consumption.
 */
export function args(strings: TemplateStringsArray): Arg[] {
  let lines = strings[0].split("\n");
  const result: Arg[] = [];
  for (let l of lines) {
    l = l.trim();
    if (l) {
      result.push(makeArg(l));
    }
  }
  return result;
}

function makeArg(str: string): Arg {
  let parts = str.match(ARG_REGEXP)!;
  let name = parts[1].trim();
  let types: ArgType[] = [];
  let isOptional = false;
  let isRepeating = false;
  for (let param of parts[2].split(",")) {
    param = param.trim().toUpperCase();
    let type = ARG_TYPES.find(t => param === t);
    if (type) {
      types.push(type);
    } else if (param === "OPTIONAL") {
      isOptional = true;
    } else if (param === "REPEATING") {
      isRepeating = true;
    }
  }
  let description = parts[3].trim();
  const result: Arg = {
    name,
    description,
    type: types
  };
  if (isOptional) {
    result.optional = true;
  }
  if (isRepeating) {
    result.repeating = true;
  }
  return result;
}
