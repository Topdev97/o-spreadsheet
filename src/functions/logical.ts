import { FunctionMap, args } from "./functions";

export const functions: FunctionMap = {
  WAIT: {
    description: "Wait",
    args: args`ms (number,cell) wait time in milliseconds`,
    returns: ["ANY"],
    async: true,
    compute: function(...args) {
      return new Promise(function(resolve, reject) {
        setTimeout(function() {
          resolve(args[0]);
        }, 2000);
      });
    }
  },
  AND: {
    description: "Returns true if all the arguments are true, false otherwise.",
    args: args`logicalPart (boolean,cell,repeating) logical part`,
    returns: ["BOOLEAN"],
    compute: function(...args) {
      return Boolean(args.flat().reduce((a, b) => a && b, true));
    }
  },
  OR: {
    description: "Returns true if at least one of the arguments are true, false otherwise.",
    args: args`logicalPart (boolean,cell,repeating) logical part`,
    returns: ["BOOLEAN"],
    compute: function(...args) {
      return Boolean(args.flat().reduce((a, b) => a || b, false));
    }
  },
  XOR: {
    description:
      "Returns true if an odd number of the provided arguments are true, false otherwise.",
    args: args`logicalPart (boolean,cell,repeating) logical part`,
    returns: ["BOOLEAN"],
    compute: function(...args) {
      return Boolean(args.flat().filter(a => a).length % 2 !== 0);
    }
  },
  NOT: {
    description: "Returns true if the argument is false, false otherwise.",
    args: args`XXX (boolean,cell) logical part`,
    returns: ["BOOLEAN"],
    compute: function(...args) {
      return Boolean(!args[0]);
    }
  },
  IF: {
    description: "Returns the second argument if the first one is true, the third otherwise.",
    args: args`
      condition (boolean,cell) logical part
      valueTrue (any) the value of the cell if the condition is true
      valueFalse (any) the value of the cell if the condition is false
    `,
    returns: ["ANY"],
    compute: function(condition, valueTrue, valueFalse) {
      return condition ? valueTrue : valueFalse;
    }
  }
};
