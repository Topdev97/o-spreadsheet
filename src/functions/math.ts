import { FunctionMap, args } from "./functions";

function toNumber(n: any) {
  return typeof n === "number" ? n : 0;
}

export const functions: FunctionMap = {
  SUM: {
    description: "Returns the sum of all values in a range.",
    compute: function(...args) {
      return args.flat().reduce((a, b) => a + toNumber(b), 0);
    },
    args: args`
        number (number,cell,range)
        numbers (number,cell,range,optional,repeating)
    `,
    returns: ["NUMBER"]
  },
  RAND: {
    description: "Returns a random number between 0 and 1",
    compute: function() {
      return Math.random();
    },
    args: [],
    returns: ["NUMBER"]
  },
  MIN: {
    description: "Returns the minimum value.",
    compute: function(...args) {
      return Math.min(...args);
    },
    args: args`
        number (number,cell,range)
        numbers (number,cell,range,optional,repeating)
    `,
    returns: ["NUMBER"]
  },
  MAX: {
    description: "Returns the maximum value.",
    compute: function(...args) {
      return Math.max(...args);
    },
    args: args`
        number (number,cell,range)
        numbers (number,cell,range,optional,repeating)
    `,
    returns: ["NUMBER"]
  }
};
