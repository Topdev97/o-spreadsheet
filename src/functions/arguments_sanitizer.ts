import { Arg } from "./arguments";

const PRIMITIVE_TYPES = ["NUMBER", "STRING", "BOOLEAN"];
const RANGE_TYPES = ["RANGE<NUMBER>", "RANGE<STRING>", "RANGE<BOOLEAN>"];

export function makeSanitizer(args: Arg[]): Function {
  const code: string[] = [];
  const hasRepeatingArg = args.length && args[args.length - 1].repeating;

  const min = args.filter(a => !a.optional).length;
  const max = args.length;
  if (min === max && !hasRepeatingArg) {
    code.push(`if (arguments.length !== ${args.length}) {`);
    code.push(
      `  throw new Error(\`Wrong number of arguments. Expected ${args.length}, but got \$\{arguments.length\} argument(s) instead.\`);`
    );
    code.push(`}`);
  } else {
    const conditions: string[] = [];
    if (min > 0) {
      conditions.push(`arguments.length < ${min}`);
    }
    if (!hasRepeatingArg) {
      conditions.push(`arguments.length > ${args.length}`);
    }
    if (conditions.length) {
      code.push(`if (${conditions.join(" || ")}) {`);
      code.push(
        `  throw new Error(\`Wrong number of arguments. Expected ${args.length}, but got \$\{arguments.length\} argument(s) instead.\`);`
      );
      code.push(`}`);
    }
  }
  const argNames: string[] = [];
  for (let i = 0; i < args.length; i++) {
    let name = `arg${i}`;
    argNames.push(args[i].repeating ? `...${name}` : `${name}`);
    sanitizeArg(code, args[i], name);
  }
  code.push(`return this.fn(${argNames.join(",")});`);
  const sanitizer = new Function(...argNames, code.join("\n"));
  return sanitizer;
}

function sanitizeArg(code: string[], arg: Arg, name: string) {
  if (arg.type.filter(t => PRIMITIVE_TYPES.includes(t)).length > 1) {
    throw new Error("Unsupported type definition. Please go talk to GED about this");
  }
  if (arg.repeating) {
    code.push(`for (let i = 0; i < ${name}.length; i++) {`);
    const basicArg = Object.assign({}, arg, { repeating: false, optional: false });
    code.push(`let ${name}_value = ${name}[i];`);
    sanitizeArg(code, basicArg, name + "_value");
    code.push(`${name}[i] = ${name}_value;`);
    code.push(`}`);
    return;
  }
  const rangeType = arg.type.find(t => RANGE_TYPES.includes(t));
  const type = rangeType ? rangeType.slice(6, -1).toLowerCase() : "";

  if (arg.type.includes("NUMBER")) {
    code.push(`switch (typeof ${name}) {`);
    if (!arg.optional || (arg.optional && arg.default)) {
      code.push(` case "undefined":`);
      code.push(`   ${name} = ${arg.default || 0};`);
      code.push(`   break;`);
    }
    code.push(` case "boolean":`);
    code.push(`   ${name} = ${name} ? 1 : 0;`);
    code.push(`   break;`);
    code.push(` case "string":`);
    code.push(`   if (${name}) {`);
    code.push(`     let n = Number(${name});`);
    code.push(`     if (isNaN(n)) {`);
    code.push(
      `       throw new Error(\`Argument "${arg.name}" should be a number, but "\$\{${name}\}" is a text, and cannot be coerced to a number.\`);`
    );
    code.push(`     } else {`);
    code.push(`       ${name} = n;`);
    code.push(`     }`);
    code.push(`   } else {`);
    code.push(`     ${name} = 0;`);
    code.push(`   }`);
    code.push(`   break;`);

    if (rangeType) {
      code.push(`  case "object":`);
      sanitizeRange(code, name, arg, type);
      code.push(`    break;`);
    }
    code.push(`}`);
    return;
  }

  if (arg.type.includes("BOOLEAN")) {
    code.push(`switch (typeof ${name}) {`);
    code.push(` case "undefined":`);
    code.push(`   ${name} = false;`);
    code.push(`   break;`);
    code.push(` case "number":`);
    code.push(`   ${name} = ${name} ? true : false;`);
    code.push(`   break;`);
    code.push(` case "string":`);
    code.push(`   if (${name}) {`);
    code.push(`     let uppercaseVal = ${name}.toUpperCase();`);
    code.push(`     if (uppercaseVal === "TRUE") {`);
    code.push(`       ${name} = true;`);
    code.push(`     } else if (uppercaseVal === "FALSE") {`);
    code.push(`       ${name} = false;`);
    code.push(`     } else {`);
    code.push(
      `       throw new Error(\`Argument "${arg.name}" should be a boolean, but "\$\{${name}\}" is a text, and cannot be coerced to a boolean.\`);`
    );
    code.push(`     }`);
    code.push(`   } else {`);
    /**
     * @compatibility Note: this is not the way Google Sheets behave:
     *
     * =if("", 1, 2) is evaluated to 2
     * =or("", 1) throws an error
     *
     * It is not clear (to me) why in the first expression it looks like it
     * is accepted, but not in the second.
     */
    code.push(`     ${name} = false;`);
    code.push(`   }`);
    code.push(`   break;`);
    code.push(`}`);
  }

  if (arg.type.includes("STRING")) {
    code.push(`switch (typeof ${name}) {`);
    code.push(` case "undefined":`);
    code.push(`   ${name} = "";`);
    code.push(`   break;`);
    code.push(` case "number":`);
    code.push(`   ${name} = ${name}.toString();`);
    code.push(`   break;`);
    code.push(` case "boolean":`);
    code.push(`   ${name} = ${name} ? "TRUE" : "FALSE";`);
    code.push(`   break;`);
    code.push(`}`);
  }

  if (arg.type.includes("RANGE")) {
    code.push(`if (!(${name} instanceof Array)) {`);
    code.push(`   throw new Error(\`Argument "${arg.name}" has the wrong type\`);`);
    code.push(`}`);
  }

  if (rangeType) {
    sanitizeRange(code, name, arg, type);
  }
}

function sanitizeRange(code: string[], name: string, arg: Arg, type: string) {
  code.push(`if (!(${name} instanceof Array)) {`);
  code.push(`   throw new Error(\`Argument "${arg.name}" has the wrong type\`);`);
  code.push(`}`);
  code.push(`for (let i = 0; i < ${name}.length; i++) {`);
  code.push(`  let col = ${name}[i];`);
  code.push(`  for (let j = 0; j < col.length; j++) {`);
  code.push(`    if (typeof col[j] !== "${type}") {`);
  code.push(`      col[j] = undefined;`);
  code.push(`    }`);
  code.push(`  }`);
  code.push(`}`);
}
//------------------------------------------------------------------------------
// Wrapping functions for arguments sanitization
//------------------------------------------------------------------------------
export function protectFunction(fn: Function, argList: Arg[]): Function {
  if (argList.length === 0) {
    return fn;
  }
  const sanitizer = makeSanitizer(argList);
  return sanitizer.bind({ fn });
}
