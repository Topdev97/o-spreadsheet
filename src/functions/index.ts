import { Arg, ArgType, protectFunction, validateArguments } from "./arguments";
import { functions as logical } from "./module_logical";
import { functions as math } from "./module_math";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------
export interface FunctionDescription {
  description: string;
  compute: Function;
  async?: boolean;
  category?: string;
  args: Arg[];
  returns: [ArgType];
}

export type FunctionMap = { [key: string]: FunctionDescription };

//------------------------------------------------------------------------------
// Functions
//------------------------------------------------------------------------------

// todo: make name more descriptive, and add some docstring

/**
 * Mapping from function names to descriptions
 */
export const functions: FunctionMap = {};

/**
 * Mapping from function name to the corresponding compute method
 */
export const functionMap: { [name: string]: Function } = {};

export { ArgType } from "./arguments";

importFunctions(math, "math");
importFunctions(logical, "logical");

function importFunctions(mapping: FunctionMap, category: string) {
  for (let name in mapping) {
    const descr = mapping[name];
    descr.category = descr.category || category;
    addFunction(name, descr);
  }
}

/**
 * Add a function to the internal function list.
 */
export function addFunction(name: string, descr: FunctionDescription) {
  name = name.toUpperCase();
  if (name in functionMap) {
    throw new Error(`Function ${name} already registered...`);
  }

  validateArguments(descr.args);

  functionMap[name] = protectFunction(descr.compute, descr.args);
  functions[name] = descr;
}
