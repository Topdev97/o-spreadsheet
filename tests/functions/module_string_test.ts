import { functionMap } from "../../src/functions/index";

const { CONCAT } = functionMap;

describe("string functions", () => {
  test("CONCAT", () => {
    expect(CONCAT("a", "b")).toBe("ab");
    expect(CONCAT("a", 334)).toBe("a334");
    expect(CONCAT("a", undefined)).toBe("a");
    expect(CONCAT("", undefined)).toBe("");
    expect(() => CONCAT()).toThrowErrorMatchingSnapshot();
    expect(() => CONCAT("asf")).toThrowErrorMatchingSnapshot();
    expect(() => CONCAT("a", "b", "c")).toThrowErrorMatchingSnapshot();
  });
});
