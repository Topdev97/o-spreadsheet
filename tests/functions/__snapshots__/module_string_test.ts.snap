// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string functions CONCAT 1`] = `"Wrong number of arguments. Expected 2, but got 0 argument(s) instead."`;

exports[`string functions CONCAT 2`] = `"Wrong number of arguments. Expected 2, but got 1 argument(s) instead."`;

exports[`string functions CONCAT 3`] = `"Wrong number of arguments. Expected 2, but got 3 argument(s) instead."`;
